src/pyqasm/visitor.py

```diff
@@ -2228,3 +2286,286 @@ def finalize(self, unrolled_stmts):
+                    span=definition.span,
+                )
+            
+            try:
+                physical_qubit_id = int(qubit_name[1:])
+                physical_qubits.append(physical_qubit_id)
+            except ValueError:
+                raise_qasm3_error(
+                    f"Invalid physical qubit identifier '{qubit_name}'. Expected format: $<integer>",
+                    error_node=definition,
+                    span=definition.span,
+                )
+        
+        print("--------------------------------")
+        return physical_qubits
+
+    def _validate_calibration_arguments(self, arguments: list, gate_name: str, definition: pulse_ast.CalibrationDefinition) -> None:
+        """Validate the classical arguments of a calibration definition."""
+        print("--------------------------------")
+        print("Validating calibration arguments: ", arguments)
+        print("--------------------------------")
+        seen_arg_names = set()
+        
+        for arg in arguments:
+            print("--------------------------------")
+            print("Argument: ", arg)
+            print("--------------------------------")
+            arg_name = arg.name.name if hasattr(arg.name, 'name') else str(arg.name)
+            
+            if arg_name in seen_arg_names:
+                raise_qasm3_error(
+                    f"Duplicate argument name '{arg_name}' in calibration definition for '{gate_name}'",
+                    error_node=definition,
+                    span=definition.span,
+                )
+            seen_arg_names.add(arg_name)
+            
+            # Validate argument type
+            type_name = type(arg.type).__name__
+            valid_types = {'AngleType', 'DurationType', 'FloatType', 'IntType', 'ComplexType'}
+            
+            if type_name not in valid_types:
+                raise_qasm3_error(
+                    f"Unsupported argument type '{type_name}' for argument '{arg_name}' "
+                    f"in calibration definition for '{gate_name}'",
+                    error_node=definition,
+                    span=definition.span,
+                )
+
+    def _validate_calibration_body(self, body: list, gate_name: str, definition: pulse_ast.CalibrationDefinition) -> None:
+        """Validate calibration body using EXISTING visitor methods."""
+        print("--------------------------------")
+        print("Validating calibration body: ", body)
+        print("--------------------------------")
+        # Set up scope for calibration arguments
+        self._push_scope({})
+        self._add_calibration_args_to_scope(definition.arguments)
+        
+        try:
+            print("--------------------------------")
+            print("Validating calibration body: ", body)
+            print("--------------------------------")
+            for stmt in body:
+                self._validate_calibration_statement_dispatch(stmt, gate_name, definition)
```

Comment:
Let's handle exceptions here. We can extract the type name and catch the raised exceptions instead of duplicating the raising in the dispatch method